java/src/main/kotlin/io/spine/validation/java/MessageValidationCode.kt

```diff
@@ -0,0 +1,237 @@
+    private val supportingMethods = mutableListOf<MethodSpec>()
+    private val constraints = mutableListOf<CodeBlock>()
+
+    /**
+     * Renders the message validation code into the provided [psiFile].
+     *
+     * @param psiFile A source file with the generated Java class for the [message].
+     *
+     * @throws IllegalStateException if [psiFile] doesn't contain a Java class for the [message].
+     */
+    fun render(psiFile: PsiJavaFile) {
+        message.ruleList.forEach(::generate)
+        val messageClass = psiFile.findClass(messageClassName)
+        val builderClass = messageClass.nested("Builder")
+        execute {
+            messageClass.apply {
+                implementValidatableMessage()
+                declareValidateMethod()
+                declareSupportingFields()
+                declareSupportingMethods()
+            }
+            builderClass.apply {
+                implementValidatingBuilder()
+                injectValidationIntoBuildMethod()
+                annotateBuildReturnType()
+                annotateBuildPartialReturnType()
+            }
+        }
+    }
+
+    /**
+     * Generates a Java [constraint][constraints], [supportingFields] and [supportingMethods]
+     * for the passed [rule].
+     *
+     * A single rule always generates a single Java constraint represented by a [CodeBlock].
+     * The number of supported members is not restricted. A single rule may generate zero,
+     * one, or more supporting members.
+     */
+    private fun generate(rule: Rule) {
+        val context = newContext(rule, message)
+        val generator = generatorFor(context)
+        constraints.add(generator.code())
+        supportingFields.addAll(generator.supportingFields())
+        supportingMethods.addAll(generator.supportingMethods())
+    }
+
+    private fun newContext(rule: Rule, message: CompilationMessage) =
+        GenerationContext(
+            typeSystem = typeSystem,
+            rule = rule,
+            msg = This(),
+            validatedType = messageType,
+            protoFile = message.type.file,
+            violationList = violations
+        )
+
+    private fun MessagePsiClass.implementValidatableMessage() {
+        val qualifiedName = ValidatableMessage::class.java.canonicalName
+        val reference = elementFactory.createInterfaceReference(qualifiedName)
+        implement(reference)
+    }
+
+    /**
+     * Declares `validate()` method in this [MessagePsiClass].
+     *
+     * This method does the following to address the formatting issues:
+     *
+     * 1. [constraints] are joined to [String] using [joinToString] instead of `joinByLine()`
+     * because the contained code blocks already have new lines when converted to string.
+     * 2. For the same reason, we trim the result because a trailing empty line is not needed.
+     * 3. When creating PSI method from the text, we have to use [trimMargin] and '|' symbols
```

Comment:
Please review this p.3 using any grammar checker. I personally use Grazie (as a Chrome extension).

I suspect some prepositions are missing.